[WIP] Upload and download large attachements through HTTP

Frando opened the pull request:
This PR combines #1589 and #1590, see those for previous discussion.

We want to support uploading large attachements to HTTP upload providers. The upload providers will allow to PUT binary blobs over HTTP to a random unique filename. They should support a simple token-based authentication. This URL will be added to the message body and a deltachat header, and the message will not include a file attachement. When receiving a message with the `Chat-Upload-Url` header, core will offer to download this attachement through through the FFI api. After downloading, the attachement should be exposed as a regular file attachement.

The current HTTP api (as in this PR, will be updated once changed) is as follows:

`{id}` is a random 16 byte id encoded as base32 selected by deltachat-core before the upload.

- `POST {endpoint}/{id}`, body is the file content. Return an error status if file exists.
- `GET {endpoint/{id}`, body is the file content

## Tasks

Rough steps, please add more and/or edit.

#### Upload

- [x] Allow to upload files to a HTTP upload URL if `DCC_UPLOAD_URL` environment variable is set
- [x] The actual pload happens in the smtp send job
- [x] The upload URL is added to the message body and as `Chat-Upload-Url` header
- [ ] Support authenticating to the server with tokens
- [ ] Support sending the original filename and/or mime type to the upload server
- [x] Encrypt files through PGP before uploading (symmetrically), include symmetric key as ~~a protected header `Chat-Upload-Key`~~ hash fragment of the URL that's part of the message
- [ ] Add config options for upload URL and threshold size after which to attach the upload
- [ ] Support scanning QR codes, detect that they are tokens and URLs for an upload provider, and set the related config after scanning

*Possibly follow-up features*
- [ ] Make file uploads resumable if they were cancelled in between. See [this comment](https://github.com/deltachat/deltachat-core-rust/pull/1589#issuecomment-642503241) for details.
- [ ] Investigate if it's worth to use a streaming cipher for symmetric encryption instead of PGP
- [ ]  Support multiple upload providers and allow to switch between them on demand

#### Download

- [x] Add FFI headers to get download status and progress and initiate downloads
- [x] `schedule_download` to start a download `Job` that downloads, decrypts, and saves attachements
- [ ] add status reporting/saving to download job
- [ ] decide on where to run the download (threadwise)
- [ ] Implement FFI API

#### Server
- [x] Add minimal demo upload server (in Node.js)
- [ ] Add token auth to demo upload server
- [ ] Support receiving original filename and/or mime type on upload, somehow save it, and set it on the download response headers
- [ ] Write spec for the HTTP API
- [ ] Settle path to production - likely try to implement server side with only NGINX config and simple command line script for token generation

*Possible follow-up features*
- [ ] Support resumable uploads
- [ ] Implement server in Rust with [tide](https://github.com/http-rs/tide)
- [ ] Support decoding in the browser with a wasm decrypter and the symmetric key provided in the hash part of the URL (which is not sent to the server)


## How to try it out

Start the demo server:

```
$ cd upload-server
$ npm install
$ node server.js

```
Start the repl with a file upload provider:
```
$ DCC_UPLOAD_URL=http://localhost:8080 RUST_LOG=info cargo run --example repl --features repl -- ~/deltachat-db
```

Create a contact and chat, and use `sendfile` or `sendimage` to send a message with an attachement.

Frando commented:
I added PGP encryption for uploads. From the commit message:

>   The symmetric key (passphrase) is put into the fragment part of the
>   upload URL (which is not sent to the server). 
>  The upload URL is part of the message as a protected header
>  and in the body payload. If the message is encrypted itself, the
>   symmetric key is encrypted with the message.

I think this should be safe? 

This would also allow to send the link with the fragment to someone else (also over non-delta channels) and allow them to read the file (e.g. `curl $URL | gpg -d $FRAGMENT > decrypted`). Is this intended? If not, we should hide the key. I think we have to think the intended UX for this through a bit, especially if this should be "only in delta messages" or if I forward such a message to someone else, they should be able (even if not super-easy for now) to decrypt the file.

dumblob commented:
>I think this should be safe?

Doesn't sound flawed on the first sight to me.

>This would also allow to send the link with the fragment to someone else (also over non-delta channels) and allow them to read the file. Is this intended?

I'd say it's intended.

>I think we have to think the intended UX for this through a bit, especially if this should be "only in delta messages" or if I forward such a message to someone else, they should be able (even if not super-easy for now) to decrypt the file.

:+1:

Speaking of servers - are there any thoughts how to lessen the probability of "misuse" of such large attachment servers in terms of "others" (i.e. anyone not using delta chat)? At least rate limiting to slow down the misuse?

I don't think the misuse is generally solvable issue except for having lots of these servers highly decentralized and redundancy as mentioned in https://github.com/deltachat/deltachat-core-rust/pull/1589#issuecomment-642503241 and https://github.com/deltachat/deltachat-core-rust/pull/1589#issuecomment-642624407 including having a user-editable list of these servers in Delta Chat apps.

Frando commented:
... and here comes the downloads :rocket:

* Added an "openfile" command to the repl to open attachements from messages (works unrelated to this PR)
* Added code to download and decrypt files that are referenced in the `Deltachat-Upload-Url` header (which is added for messages where the attachement was uploaded to http).
* Added an "download" command to the repl that starts a Job `DownloadMessageFile` for a message id that triggers the download.

One question I encountered was on which thread the download should run. I've put it into the Imap thread for now, but I'm not familiar with the thread architecture of deltachat - or do we want a new thread for downloads?

Status reporting and integration with the C FFI is still missing. I'm also not really familiar to neither the C FFI nor the Job API yet so if someone comments on these or wants to start the work please go ahead! (But note here in the PR so that we don't duplicate).

hpk42 commented:
haven't looked deeply into it yet but there is an XMPP "http-file-upload" standard https://xmpp.org/extensions/xep-0363.html to which @inputmice contributed. sidenote: He is also involved with "lttr.rs" an android app using JMAP for e-mail protocol.  Unfortunately the ".rs" does not mean it involves Rust :)

r10s commented:
moving stale pr to [resurrection](https://github.com/orgs/deltachat/projects/19) due to [march2022 cleanup](https://github.com/orgs/deltachat/projects/48).
